src/Interpreter.ts

```diff
@@ -212,95 +228,38 @@ export default class Interpreter {
         contestOffset + matchedTemplate.contests.length
       )
     )
-    const votes = this.getVotesFromMarks(
-      getContests({ ballotStyle, election }).slice(
-        contestOffset,
-        contestOffset + matchedTemplate.contests.length
-      ),
-      marks
-    )
 
-    const ballot: CompletedBallot = {
+    return { matchedTemplate, metadata, marks }
+  }
+
+  private interpretMarks({
+    marks,
+    metadata,
+  }: FindMarksResult): CompletedBallot {
+    const { election } = this
+    const ballotStyle = defined(
+      getBallotStyle({
+        ballotStyleId: metadata.ballotStyleId,
+        election,
+      })
+    )
+    const precinct = defined(
+      getPrecinctById({
+        election,
+        precinctId: metadata.precinctId,
+      })
+    )
+    return {
       ballotId: uuid(),
       ballotStyle,
       ballotType: BallotType.Standard,
-      election: this.election,
-      isTestBallot: false,
+      election,
+      isTestBallot: metadata.isTestBallot,
       precinct,
-      votes,
-    }
-
-    return { matchedTemplate, ballot, marks }
-  }
-
-  private getVotesFromMarks(
```

Comment:
All this became much more simplified since each `mark` object includes the information about contest and candidate etc. It was moved into its own file, but is much smaller now.